README
cUnidecode, lossy ASCII transliterations of Unicode text
=======================================================

It often happens that you have text data in Unicode, but you need to
represent it in ASCII. For example when integrating with legacy code that
doesn't support Unicode, or for ease of entry of non-Roman names on a US
keyboard, or when constructing ASCII machine identifiers from
human-readable Unicode strings that should still be somewhat intelligible
(a popular example of this is when making an URL slug from an article
title). 

In most of these examples you could represent Unicode characters as
"???" or "\\15BA\\15A0\\1610", to mention two extreme cases. But that's
nearly useless to someone who actually wants to read what the text says.

What Unidecode provides is a middle road: function unidecode() takes
Unicode data and tries to represent it in ASCII characters (i.e., the
universally displayable characters between 0x00 and 0x7F), where the
compromises taken when mapping between two character sets are chosen to be
near what a human with a US keyboard would choose.

The quality of resulting ASCII representation varies. For languages of
western origin it should be between perfect and good. On the other hand
transliteration (i.e., conveying, in Roman letters, the pronunciation
expressed by the text in some other writing system) of languages like
Chinese, Japanese or Korean is a very complex issue and this library does
not even attempt to address it. It draws the line at context-free
character-by-character mapping. So a good rule of thumb is that the further
the script you are transliterating is from Latin alphabet, the worse the
transliteration will be.

Note that this module generally produces better results than simply
stripping accents from characters (which can be done in Python with
built-in functions). It is based on hand-tuned character mappings that for
example also contain ASCII approximations for symbols and non-Latin
alphabets.

This is a C port of Unidecode by Paul Logston <code@logston.me>

Unidecode (by Tomaz Solc) is a Python port of the Text::Unidecode Perl module by
Sean M. Burke <sburke@cpan.org> and later


Module content
--------------

The module exports a single function that takes an Unicode object (Python
2.7) and returns a string::

    >>> from cunidecode import unidecode
    >>> unidecode(u'ko\u017eu\u0161\u010dek')
    'kozuscek'
    >>> unidecode(u'30 \U0001d5c4\U0001d5c6/\U0001d5c1')
    '30 km/h'
    >>> unidecode(u"\u5317\u4EB0")
    'Bei Jing '


Requirements
------------

Python 2.7 and a C compiler/linker.
   
This implementation of unidecode does not "decode" characters outside
of the Basic Multilingual Plane.
A Python build with "wide" Unicode characters may lead to a segmentation
fault if an attempt to decode a character outside the BMP is made.

Surrogate pair encoding of "narrow" builds is not supported.


Installation
------------

You install Unidecode, as you would install any Python module, by running
these commands::

    python setup.py install
    python setup.py test


Source
------

You can get the latest development version of cUnidecode with::

    git clone git@github.com:logston/cunidecode.git


Support
-------

Questions, bug reports, useful code bits, and suggestions for Unidecode
should be sent to tomaz.solc@tablix.org

Questions, bug reports, useful code bits, and suggestions for cUnidecode
should be sent to code@logston.me


Copyright
---------

Original character transliteration tables:

Copyright 2001, Sean M. Burke <sburke@cpan.org>, all rights reserved.

Python code and later additions to Unidecode:

Copyright 2014, Tomaz Solc <tomaz.solc@tablix.org>

C code and later additions to cUnidecode:

Copyright 2014, Paul Logston <code@logston.me>

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation; either version 2 of the License, or (at your option)
any later version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc., 51
Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.  The programs and
documentation in this dist are distributed in the hope that they will be
useful, but without any warranty; without even the implied warranty of
merchantability or fitness for a particular purpose.



ChangeLog
2014-08-25 cunidecode 0.0.1

	* Update README for added clarity.

2014-08-22 cunidecode 0.0.1

	* Built C port to unidecode

2014-05-11	unidecode 0.04.16

	* Corrected ChangeLog.

2014-05-11	unidecode 0.04.15

	* Fixed transliteration of Chinese character "Yi". (thanks to Yao Zuo)
	* Transliterate Euro sign as "EUR". (thanks to Dave Smith)

2013-09-20	unidecode 0.04.14

	* Revert transliteration for Latin characters with umlauts back
	  to language-neutral, unaccented characters.
	* Added transliterations for circled numbers, letters and ordinals.
	* Minor fixes to rarely used punctuation characters.

2013-05-30	unidecode 0.04.13

	* Use German transliteration for Latin characters with umlauts.
	  (thanks to Geoffrey Teale)
	* Fixed transliteration for several rarely used punctuation
	  characters.
	* Add transliteration for trademark and similar signs. (thanks to
	  Ovanes Markaryan)

2013-01-28	unidecode 0.04.12

	* Tests now pass on Python 2.7.
	* Make argument type warning refer to unidecode()'s caller.

2013-01-17	unidecode 0.04.11

	* Issue a warning on Python 2.x if an object that is not an instance of
	  unicode class is passed to unidecode().
	* Fixed unit tests for Python 2.5.

2012-12-01	unidecode 0.04.10

	* Added trove classifiers.

2011-09-23	unidecode 0.04.9

	* Fixed Python 2.5 installation issue.

2011-09-22	unidecode 0.04.8

	* License change to GPL to fit better with other Python modules and
	  to remove the issue of depending on Perl for licensing terms.
	* Always return a string object on Python 2.x. Before this patch,
	  unidecode() returned a unicode object on Python 2.x if the input
	  was a unicode object that contained ASCII characters. Behaviour
	  on Python 3.x remains unchanged. (thanks to Wesley Yarde)

2011-04-04	unidecode 0.04.7

	* Unit test fixes for Python builds using 16-bit Unicode
	  representation (see PEP 261)

2011-02-16	unidecode 0.04.6

	* Support for a lot of various less common variants of Latin
	  letters and digraphs.
	* Fixes to Czech pangram transliteration (thanks to Jan Killian)

2010-10-15	unidecode 0.04.5

	* Unit test fixes.

2010-10-12	unidecode 0.04.4

	* Unit tests for Python 3 and simplified transliteration
	  code (thanks to Thomas Kluyver)
	* Support for mathematical versions of Latin and Greek alphabets 
	  and digits.

2010-01-12	unidecode 0.04.3

	* Corrected transliteration for Cyrillic alphabet
	  (thanks to Ruslan Grokhovetskiy)

2009-08-27	unidecode 0.04.2

	* Corrected replacements for Japanese Kana characters 
	  (thanks to Michel Albert)

2009-01-25	unidecode 0.04.1

	* First public release


tests/test_basic_2.py
# -*- coding: utf-8 -*-
import unittest
import sys
from cunidecode import unidecode
import warnings

# workaround for Python < 2.7
if not hasattr(unittest, 'skipIf'):
    def skipIf(condition, reason):
        def d(f):
            def df(*args):
                if condition:
                    print ("skipped %r" % (reason,))
                else:
                    return f(*args)
            return df
        return d
    unittest.skipIf = skipIf

class TestUnidecode(unittest.TestCase):
    def test_ascii(self):
        for n in xrange(32,128):
            t = unichr(n)
            self.assertEqual(unidecode(t), t)

    def test_bmp(self):
        for n in range(0, 0x10000):
            # Just check that it doesn't throw an exception
            t = unichr(n)
            unidecode(t)

    def test_circled_latin(self):
        # 1 sequence of a-z
        for n in range(0, 26):
            a = chr(ord('a') + n)
            b = unidecode(unichr(0x24d0 + n))

            self.assertEqual(b, a)

    def test_specific(self):

        TESTS = [
 
                (u"Hello, World!", 
                "Hello, World!"),

                (u"'\"\r\n",
                 "'\"\r\n"),

                (u"ČŽŠčžš",
                 "CZSczs"),

                (u"ア",
                 "a"),

                (u"α",
                "a"),

                (u"а",
                "a"),

                (u'ch\xe2teau',
                "chateau"),

                (u'vi\xf1edos',
                "vinedos"),
                
                (u"\u5317\u4EB0",
                "Bei Jing "),

                (u"Efﬁcient",
                "Efficient"),

                # https://github.com/iki/unidecode/commit/4a1d4e0a7b5a11796dc701099556876e7a520065
                (u'příliš žluťoučký kůň pěl ďábelské ódy',
                'prilis zlutoucky kun pel dabelske ody'),

                (u'PŘÍLIŠ ŽLUŤOUČKÝ KŮŇ PĚL ĎÁBELSKÉ ÓDY',
                'PRILIS ZLUTOUCKY KUN PEL DABELSKE ODY'),

                # Table that doesn't exist
                (u'\ua500',
                ''),
                
                # Table that has less than 256 entriees
                (u'\u1eff',
                ''),
            ]

        for input, correct_output in TESTS:
            test_output = unidecode(input)
            self.assertEqual(test_output, correct_output)
            self.assertTrue(isinstance(test_output, str))

    def test_wordpress_remove_accents(self):
        # This is the table from remove_accents() WordPress function.
        # https://core.trac.wordpress.org/browser/trunk/wp-includes/formatting.php

        wp_remove_accents = {
            # Decompositions for Latin-1 Supplement
            chr(194)+chr(170) : 'a', chr(194)+chr(186) : 'o',
            chr(195)+chr(128) : 'A', chr(195)+chr(129) : 'A',
            chr(195)+chr(130) : 'A', chr(195)+chr(131) : 'A',
            chr(195)+chr(133) : 'A',
            chr(195)+chr(134) : 'AE',chr(195)+chr(135) : 'C',
            chr(195)+chr(136) : 'E', chr(195)+chr(137) : 'E',
            chr(195)+chr(138) : 'E', chr(195)+chr(139) : 'E',
            chr(195)+chr(140) : 'I', chr(195)+chr(141) : 'I',
            chr(195)+chr(142) : 'I', chr(195)+chr(143) : 'I',
            chr(195)+chr(144) : 'D', chr(195)+chr(145) : 'N',
            chr(195)+chr(146) : 'O', chr(195)+chr(147) : 'O',
            chr(195)+chr(148) : 'O', chr(195)+chr(149) : 'O',
            chr(195)+chr(153) : 'U',
            chr(195)+chr(154) : 'U', chr(195)+chr(155) : 'U',
            chr(195)+chr(157) : 'Y',
            chr(195)+chr(160) : 'a', chr(195)+chr(161) : 'a',
            chr(195)+chr(162) : 'a', chr(195)+chr(163) : 'a',
            chr(195)+chr(165) : 'a',
            chr(195)+chr(166) : 'ae',chr(195)+chr(167) : 'c',
            chr(195)+chr(168) : 'e', chr(195)+chr(169) : 'e',
            chr(195)+chr(170) : 'e', chr(195)+chr(171) : 'e',
            chr(195)+chr(172) : 'i', chr(195)+chr(173) : 'i',
            chr(195)+chr(174) : 'i', chr(195)+chr(175) : 'i',
            chr(195)+chr(176) : 'd', chr(195)+chr(177) : 'n',
            chr(195)+chr(178) : 'o', chr(195)+chr(179) : 'o',
            chr(195)+chr(180) : 'o', chr(195)+chr(181) : 'o',
            chr(195)+chr(184) : 'o',
            chr(195)+chr(185) : 'u', chr(195)+chr(186) : 'u',
            chr(195)+chr(187) : 'u',
            chr(195)+chr(189) : 'y', chr(195)+chr(190) : 'th',
            chr(195)+chr(191) : 'y', chr(195)+chr(152) : 'O',
            # Decompositions for Latin Extended-A
            chr(196)+chr(128) : 'A', chr(196)+chr(129) : 'a',
            chr(196)+chr(130) : 'A', chr(196)+chr(131) : 'a',
            chr(196)+chr(132) : 'A', chr(196)+chr(133) : 'a',
            chr(196)+chr(134) : 'C', chr(196)+chr(135) : 'c',
            chr(196)+chr(136) : 'C', chr(196)+chr(137) : 'c',
            chr(196)+chr(138) : 'C', chr(196)+chr(139) : 'c',
            chr(196)+chr(140) : 'C', chr(196)+chr(141) : 'c',
            chr(196)+chr(142) : 'D', chr(196)+chr(143) : 'd',
            chr(196)+chr(144) : 'D', chr(196)+chr(145) : 'd',
            chr(196)+chr(146) : 'E', chr(196)+chr(147) : 'e',
            chr(196)+chr(148) : 'E', chr(196)+chr(149) : 'e',
            chr(196)+chr(150) : 'E', chr(196)+chr(151) : 'e',
            chr(196)+chr(152) : 'E', chr(196)+chr(153) : 'e',
            chr(196)+chr(154) : 'E', chr(196)+chr(155) : 'e',
            chr(196)+chr(156) : 'G', chr(196)+chr(157) : 'g',
            chr(196)+chr(158) : 'G', chr(196)+chr(159) : 'g',
            chr(196)+chr(160) : 'G', chr(196)+chr(161) : 'g',
            chr(196)+chr(162) : 'G', chr(196)+chr(163) : 'g',
            chr(196)+chr(164) : 'H', chr(196)+chr(165) : 'h',
            chr(196)+chr(166) : 'H', chr(196)+chr(167) : 'h',
            chr(196)+chr(168) : 'I', chr(196)+chr(169) : 'i',
            chr(196)+chr(170) : 'I', chr(196)+chr(171) : 'i',
            chr(196)+chr(172) : 'I', chr(196)+chr(173) : 'i',
            chr(196)+chr(174) : 'I', chr(196)+chr(175) : 'i',
            chr(196)+chr(176) : 'I', chr(196)+chr(177) : 'i',
            chr(196)+chr(178) : 'IJ',chr(196)+chr(179) : 'ij',
            chr(196)+chr(180) : 'J', chr(196)+chr(181) : 'j',
            chr(196)+chr(182) : 'K', chr(196)+chr(183) : 'k',
            chr(196)+chr(184) : 'k', chr(196)+chr(185) : 'L',
            chr(196)+chr(186) : 'l', chr(196)+chr(187) : 'L',
            chr(196)+chr(188) : 'l', chr(196)+chr(189) : 'L',
            chr(196)+chr(190) : 'l', chr(196)+chr(191) : 'L',
            chr(197)+chr(128) : 'l', chr(197)+chr(129) : 'L',
            chr(197)+chr(130) : 'l', chr(197)+chr(131) : 'N',
            chr(197)+chr(132) : 'n', chr(197)+chr(133) : 'N',
            chr(197)+chr(134) : 'n', chr(197)+chr(135) : 'N',
            chr(197)+chr(136) : 'n',
            chr(197)+chr(140) : 'O', chr(197)+chr(141) : 'o',
            chr(197)+chr(142) : 'O', chr(197)+chr(143) : 'o',
            chr(197)+chr(144) : 'O', chr(197)+chr(145) : 'o',
            chr(197)+chr(146) : 'OE',chr(197)+chr(147) : 'oe',
            chr(197)+chr(148) : 'R',chr(197)+chr(149) : 'r',
            chr(197)+chr(150) : 'R',chr(197)+chr(151) : 'r',
            chr(197)+chr(152) : 'R',chr(197)+chr(153) : 'r',
            chr(197)+chr(154) : 'S',chr(197)+chr(155) : 's',
            chr(197)+chr(156) : 'S',chr(197)+chr(157) : 's',
            chr(197)+chr(158) : 'S',chr(197)+chr(159) : 's',
            chr(197)+chr(160) : 'S', chr(197)+chr(161) : 's',
            chr(197)+chr(162) : 'T', chr(197)+chr(163) : 't',
            chr(197)+chr(164) : 'T', chr(197)+chr(165) : 't',
            chr(197)+chr(166) : 'T', chr(197)+chr(167) : 't',
            chr(197)+chr(168) : 'U', chr(197)+chr(169) : 'u',
            chr(197)+chr(170) : 'U', chr(197)+chr(171) : 'u',
            chr(197)+chr(172) : 'U', chr(197)+chr(173) : 'u',
            chr(197)+chr(174) : 'U', chr(197)+chr(175) : 'u',
            chr(197)+chr(176) : 'U', chr(197)+chr(177) : 'u',
            chr(197)+chr(178) : 'U', chr(197)+chr(179) : 'u',
            chr(197)+chr(180) : 'W', chr(197)+chr(181) : 'w',
            chr(197)+chr(182) : 'Y', chr(197)+chr(183) : 'y',
            chr(197)+chr(184) : 'Y', chr(197)+chr(185) : 'Z',
            chr(197)+chr(186) : 'z', chr(197)+chr(187) : 'Z',
            chr(197)+chr(188) : 'z', chr(197)+chr(189) : 'Z',
            chr(197)+chr(190) : 'z', chr(197)+chr(191) : 's',
            # Decompositions for Latin Extended-B
            chr(200)+chr(152) : 'S', chr(200)+chr(153) : 's',
            chr(200)+chr(154) : 'T', chr(200)+chr(155) : 't',

            # Vowels with diacritic (Vietnamese)
            # unmarked
            chr(198)+chr(160) : 'O', chr(198)+chr(161) : 'o',
            chr(198)+chr(175) : 'U', chr(198)+chr(176) : 'u',
            # grave accent
            chr(225)+chr(186)+chr(166) : 'A', chr(225)+chr(186)+chr(167) : 'a',
            chr(225)+chr(186)+chr(176) : 'A', chr(225)+chr(186)+chr(177) : 'a',
            chr(225)+chr(187)+chr(128) : 'E', chr(225)+chr(187)+chr(129) : 'e',
            chr(225)+chr(187)+chr(146) : 'O', chr(225)+chr(187)+chr(147) : 'o',
            chr(225)+chr(187)+chr(156) : 'O', chr(225)+chr(187)+chr(157) : 'o',
            chr(225)+chr(187)+chr(170) : 'U', chr(225)+chr(187)+chr(171) : 'u',
            chr(225)+chr(187)+chr(178) : 'Y', chr(225)+chr(187)+chr(179) : 'y',
            # hook
            chr(225)+chr(186)+chr(162) : 'A', chr(225)+chr(186)+chr(163) : 'a',
            chr(225)+chr(186)+chr(168) : 'A', chr(225)+chr(186)+chr(169) : 'a',
            chr(225)+chr(186)+chr(178) : 'A', chr(225)+chr(186)+chr(179) : 'a',
            chr(225)+chr(186)+chr(186) : 'E', chr(225)+chr(186)+chr(187) : 'e',
            chr(225)+chr(187)+chr(130) : 'E', chr(225)+chr(187)+chr(131) : 'e',
            chr(225)+chr(187)+chr(136) : 'I', chr(225)+chr(187)+chr(137) : 'i',
            chr(225)+chr(187)+chr(142) : 'O', chr(225)+chr(187)+chr(143) : 'o',
            chr(225)+chr(187)+chr(148) : 'O', chr(225)+chr(187)+chr(149) : 'o',
            chr(225)+chr(187)+chr(158) : 'O', chr(225)+chr(187)+chr(159) : 'o',
            chr(225)+chr(187)+chr(166) : 'U', chr(225)+chr(187)+chr(167) : 'u',
            chr(225)+chr(187)+chr(172) : 'U', chr(225)+chr(187)+chr(173) : 'u',
            chr(225)+chr(187)+chr(182) : 'Y', chr(225)+chr(187)+chr(183) : 'y',
            # tilde
            chr(225)+chr(186)+chr(170) : 'A', chr(225)+chr(186)+chr(171) : 'a',
            chr(225)+chr(186)+chr(180) : 'A', chr(225)+chr(186)+chr(181) : 'a',
            chr(225)+chr(186)+chr(188) : 'E', chr(225)+chr(186)+chr(189) : 'e',
            chr(225)+chr(187)+chr(132) : 'E', chr(225)+chr(187)+chr(133) : 'e',
            chr(225)+chr(187)+chr(150) : 'O', chr(225)+chr(187)+chr(151) : 'o',
            chr(225)+chr(187)+chr(160) : 'O', chr(225)+chr(187)+chr(161) : 'o',
            chr(225)+chr(187)+chr(174) : 'U', chr(225)+chr(187)+chr(175) : 'u',
            chr(225)+chr(187)+chr(184) : 'Y', chr(225)+chr(187)+chr(185) : 'y',
            # acute accent
            chr(225)+chr(186)+chr(164) : 'A', chr(225)+chr(186)+chr(165) : 'a',
            chr(225)+chr(186)+chr(174) : 'A', chr(225)+chr(186)+chr(175) : 'a',
            chr(225)+chr(186)+chr(190) : 'E', chr(225)+chr(186)+chr(191) : 'e',
            chr(225)+chr(187)+chr(144) : 'O', chr(225)+chr(187)+chr(145) : 'o',
            chr(225)+chr(187)+chr(154) : 'O', chr(225)+chr(187)+chr(155) : 'o',
            chr(225)+chr(187)+chr(168) : 'U', chr(225)+chr(187)+chr(169) : 'u',
            # dot below
            chr(225)+chr(186)+chr(160) : 'A', chr(225)+chr(186)+chr(161) : 'a',
            chr(225)+chr(186)+chr(172) : 'A', chr(225)+chr(186)+chr(173) : 'a',
            chr(225)+chr(186)+chr(182) : 'A', chr(225)+chr(186)+chr(183) : 'a',
            chr(225)+chr(186)+chr(184) : 'E', chr(225)+chr(186)+chr(185) : 'e',
            chr(225)+chr(187)+chr(134) : 'E', chr(225)+chr(187)+chr(135) : 'e',
            chr(225)+chr(187)+chr(138) : 'I', chr(225)+chr(187)+chr(139) : 'i',
            chr(225)+chr(187)+chr(140) : 'O', chr(225)+chr(187)+chr(141) : 'o',
            chr(225)+chr(187)+chr(152) : 'O', chr(225)+chr(187)+chr(153) : 'o',
            chr(225)+chr(187)+chr(162) : 'O', chr(225)+chr(187)+chr(163) : 'o',
            chr(225)+chr(187)+chr(164) : 'U', chr(225)+chr(187)+chr(165) : 'u',
            chr(225)+chr(187)+chr(176) : 'U', chr(225)+chr(187)+chr(177) : 'u',
            chr(225)+chr(187)+chr(180) : 'Y', chr(225)+chr(187)+chr(181) : 'y',
            # Vowels with diacritic (Chinese, Hanyu Pinyin)
            chr(201)+chr(145) : 'a',
            # macron
            chr(199)+chr(149) : 'U', chr(199)+chr(150) : 'u',
            # acute accent
            chr(199)+chr(151) : 'U', chr(199)+chr(152) : 'u',
            # caron
            chr(199)+chr(141) : 'A', chr(199)+chr(142) : 'a',
            chr(199)+chr(143) : 'I', chr(199)+chr(144) : 'i',
            chr(199)+chr(145) : 'O', chr(199)+chr(146) : 'o',
            chr(199)+chr(147) : 'U', chr(199)+chr(148) : 'u',
            chr(199)+chr(153) : 'U', chr(199)+chr(154) : 'u',
            # grave accent
            chr(199)+chr(155) : 'U', chr(199)+chr(156) : 'u',

            chr(195)+chr(132) : 'A',
            chr(195)+chr(150) : 'O',
            chr(195)+chr(156) : 'U',
            #chr(195)+chr(159) : 's',
            chr(195)+chr(164) : 'a',
            chr(195)+chr(182) : 'o',
            chr(195)+chr(188) : 'u',

            # Known differences:

            #chr(195)+chr(158) : 'TH',
            #chr(197)+chr(137) : 'N',
            #chr(197)+chr(138) : 'n',
            #chr(197)+chr(139) : 'N',

            # Euro Sign
            #chr(226)+chr(130)+chr(172) : 'E',

            # GBP (Pound) Sign
            #chr(194)+chr(163) : '',
        }

        for utf8_input, correct_output in wp_remove_accents.iteritems():
            input = utf8_input.decode('utf8')
            output = unidecode(input)

            self.assertEqual(correct_output, output)

if __name__ == "__main__":
    unittest.main()


unidecode.c
#include <Python.h>
#include "data.h"




static PyObject *unidecode_unidecode( PyObject *self, PyObject *args ) {
    Py_UNICODE *string;
    int string_size;

    if (!PyArg_ParseTuple(args, "u#", &string, &string_size)) {
      return NULL;
    } 

    int c, section, position;
    for (int i=0; i < string_size; i++) {
        c = string[i];
        section = c >> 8;
        position = c % 256;

        char* s = data[section][position];
    }

    return Py_BuildValue("(si)", string, string_size);
}

static PyMethodDef unidecode_methods[] = {
    { "unidecode", (PyCFunction)unidecode_unidecode, METH_VARARGS,
      "Transliterate an Unicode object into an ASCII string" },
    { NULL, NULL, 0, NULL }
};


char* module_doc = "\
Transliterate Unicode text into plain 7-bit ASCII.\
\
Example usage:\
>>> from unidecode import unidecode:\
>>> unidecode(u'\u5317\u4EB0')\
'Bei Jing'\
\
The transliteration uses a straightforward map, and doesn't have alternatives\
for the same character based on language, position, or anything else.\
\
In Python 3, a standard string object will be returned. If you need bytes, use:\
>>> unidecode('Κνωσός').encode('ascii')\
b'Knosos'\
";

PyMODINIT_FUNC initunidecode() {
    Py_InitModule3("unidecode", unidecode_methods, module_doc);
}



cunidecode.c
#include <stdio.h>
#include <string.h>
#include <Python.h>
#include "data.h"


static int get_string_size(char *string) {
    int size = 0;
    while (string[size] != '\0') {
        size++;
    }
    return size;
}


static char *append(char *str_head, char *str_tail) { 
    int head_size = get_string_size(str_head);
    int tail_size = get_string_size(str_tail);
    int total_size = head_size + tail_size;

    char *ret_string = (char *)malloc(sizeof(char) * (total_size + 1));

    strncpy(ret_string, str_head, head_size);
    strncpy((ret_string + head_size), str_tail, tail_size);
    ret_string[total_size] = '\0';

    free(str_head);

    return ret_string;
}


static PyObject *cunidecode_unidecode( PyObject *self, PyObject *args ) {
    Py_UNICODE *string;
    int string_size;
    if (!PyArg_ParseTuple(args, "u#", &string, &string_size)) {
      return NULL;
    } 

    char *temp_string;
    // Build an initial buffer the size of the unicode string.
    char *ret_string = (char *)malloc(sizeof(char));
    if (ret_string == 0) {
      EXIT_FAILURE;
    }
      
    ret_string[0] = '\0';

    int i, unichar, section, position;
    for (i = 0; i < string_size; i++) {
        unichar = string[i];

        // Only support the Basic Multilingual Plane
        if (unichar < 65536) {
            section = unichar >> 8;
            position = unichar % 256;
            temp_string = data[section][position];
        } else {
            temp_string = "";
        }

        ret_string = append(ret_string, temp_string);
    }

    PyObject* ret_val = Py_BuildValue("s", ret_string);

    free(ret_string);

    return ret_val;
}


static PyMethodDef cunidecode_methods[] = {
    { "unidecode", (PyCFunction)cunidecode_unidecode, METH_VARARGS,
      "Transliterate an Unicode object into an ASCII string" },
    { NULL, NULL, 0, NULL }
};


char* module_doc = "\
Transliterate Unicode text into plain 7-bit ASCII.\
\
Example usage:\
>>> from cunidecode import unidecode:\
>>> unidecode(u'\u5317\u4EB0')\
'Bei Jing'\
\
The transliteration uses a straightforward map, and doesn't have alternatives\
for the same character based on language, position, or anything else.\
";

PyMODINIT_FUNC initcunidecode() {
    Py_InitModule3("cunidecode", cunidecode_methods, module_doc);
}



convert_data.py
from collections import OrderedDict
import glob
import os
from cStringIO import StringIO

CUR_DIR = os.path.dirname(os.path.abspath(__file__))
DATA_DIR = os.path.join(CUR_DIR, 'data')


def escape(string):

    if '\\' in string:
        string = string.replace('\\', '\\\\')

    if '\n' in string:
        string = string.replace('\n', '\\\n')

    if '\r' in string:
        string = string.replace('\r', '\\\r')

    if '"' in string:
        string = string.replace('"', '\\"')

    return string


def convert_data():
    file_name_pattern = 'x0*.py'
    pattern = os.path.join(DATA_DIR, file_name_pattern)
    data_dict = OrderedDict((i, None) for i in xrange(256))
    for file in glob.iglob(pattern):
        mod_hex_name = os.path.basename(file)[:-3]
        mod_dec_name = int('0' + mod_hex_name, base=16)
        mod = __import__('data.%s' % mod_hex_name, [], [], ['data'])
        data = list(mod.data)
        if len(data) != 256:
            diff = 256 - len(data)
            data.extend(['' for _ in xrange(diff)])
        data_dict[mod_dec_name] = data

    for k, v in data_dict.items():
        if v is None:
            data_dict[k] = ['' for _ in xrange(256)]

    # Only transliterate two characters below ASCII code point 32
    data_dict[0][10] = '\n'
    data_dict[0][13] = '\r'

    for i in xrange(32, 128):
        data_dict[0][i] = chr(i)

    with open(os.path.join(CUR_DIR, 'data.h'), 'w') as s:
        s.write('char* data[256][256] = {\n')

        for strings in data_dict.values():
            s.write('\t{\n')

            string_list = []
            for i, string in enumerate(strings):
                if string == '\n':  # Treat \n as a special case
                    string_list.append('\t\t"\\n",')
                elif string == '\r':  # Treat \r as a special case
                    string_list.append('\t\t"\\r",')
                else:
                    string_list.append(
                        '\t\t"%s"%s' % (escape(string), '' if i == 255 else ',')
                    )

            s.write('\n'.join(string_list))

            s.write('\n\t},\n')

        s.write('};\n')


if __name__ == '__main__':
    convert_data()
